plugin/storage/es/spanstore/reader.go

```diff
@@ -217,6 +255,173 @@ func (s *SpanReader) bucketToStringArray(buckets []*elastic.AggregationBucketKey
 
 // FindTraces retrieves traces that match the traceQuery
 func (s *SpanReader) FindTraces(traceQuery *spanstore.TraceQueryParameters) ([]*model.Trace, error) {
-	// TODO
-	return nil, nil
+	if err := validateQuery(traceQuery); err != nil {
+		return nil, err
+	}
+	if traceQuery.NumTraces == 0 {
+		traceQuery.NumTraces = defaultNumTraces
+	}
+	uniqueTraceIDs, err := s.findTraceIDs(traceQuery)
+	if err != nil {
+		return nil, err
+	}
+	var retMe []*model.Trace
+	for traceID := range uniqueTraceIDs {
+		if len(retMe) >= traceQuery.NumTraces {
+			break
+		}
+		trace, err := s.readTrace(string(traceID), traceQuery)
+		if err != nil {
+			s.logger.Error("Failure to read trace", zap.String("trace_id", string(traceID)), zap.Error(err))
+			continue
+		}
+		retMe = append(retMe, trace)
+	}
+	return retMe, nil
+}
+
+func validateQuery(p *spanstore.TraceQueryParameters) error {
+	if p == nil {
+		return ErrMalformedRequestObject
+	}
+	if p.ServiceName == "" && len(p.Tags) > 0 {
+		return ErrServiceNameNotSet
+	}
+	if p.StartTimeMin.IsZero() || p.StartTimeMax.IsZero() {
+		return ErrStartAndEndTimeNotSet
+	}
+	if !p.StartTimeMin.IsZero() && !p.StartTimeMax.IsZero() && p.StartTimeMax.Before(p.StartTimeMin) {
+		return ErrStartTimeMinGreaterThanMax
+	}
+	if p.DurationMin != 0 && p.DurationMax != 0 && p.DurationMin > p.DurationMax {
+		return ErrDurationMinGreaterThanMax
+	}
+	if (p.DurationMin != 0 || p.DurationMax != 0) && len(p.Tags) > 0 {
+		return ErrDurationAndTagQueryNotSupported
+	}
+	return nil
+}
+
+func (s *SpanReader) findTraceIDs(traceQuery *spanstore.TraceQueryParameters) ([]string, error) {
+	//  Below is the JSON body to our HTTP GET request to ElasticSearch. This function creates this.
+	//
+	// {
+	//     "size": 0,
+	//     "query": {
+	//       "bool": {
+	//         "must": [
+	//           { "match": { "operationName":   "traceQuery.OperationName"      }},
+	//           { "match": { "process.serviceName": "traceQuery.ServiceName" }},
+	//           { "range":  { "timestamp": { "gte": traceQuery.StartTimeMin, "lte": traceQuery.StartTimeMax }}},
+	//           { "range":  { "duration": { "gte": traceQuery.DurationMin, "lte": traceQuery.DurationMax }}},
+	//           { "nested" : {
+	//             "path" : "tags",
```

Comment:
No it doesn't but it's a little more tricky than i thought it would be, so i left it as a todo for later

i didnt write that anywhere did i.. i'll do that next commit